helix-term/src/ui/editor.rs

```diff
@@ -695,6 +750,343 @@ impl EditorView {
+                        helix_core::LineEnding::Crlf => 2,
+                        helix_core::LineEnding::LF => 1,
+                    });
+
+                surface.set_stringn(
+                    new_x_location,
+                    additional_area.y,
+                    dots,
+                    dots.len(),
+                    theme.get("keyword.operator"),
+                );
+            }
+
+            new_offset.anchor = text.byte_to_char(node.byte_range.start);
+
+            // get all highlights from the latest point
+            let highlights = Self::doc_syntax_highlights(doc, new_offset.anchor, 1, theme);
+
+            let mut renderer = TextRenderer::new(
+                surface,
+                doc,
+                theme,
+                view.offset.horizontal_offset,
+                line_context_area,
+            );
+
+            // limit the width to its size - 1, so that it won't draw trailing whitespace characters
+            line_context_area.width = already_written - 1;
+
+            render_text(
+                &mut renderer,
+                text,
+                new_offset,
+                &doc.text_format(line_context_area.width, Some(theme)),
+                &virtual_text_annotations,
+                highlights,
+                theme,
+                line_decoration,
+                translated_positions,
+            );
+
+            // next node
+            context_area.y += 1;
+        }
+    }
+
+    fn get_context_paired_range(
+        query_match: &QueryMatch,
+        start_index: u32,
+        end_index: u32,
+        top_first_byte: usize,
+        last_scan_byte: usize,
+    ) -> Option<std::ops::Range<usize>> {
+        let end_nodes: Vec<_> = query_match.nodes_for_capture_index(end_index).collect();
+        query_match
+            .nodes_for_capture_index(start_index)
+            .flat_map(move |context| {
+                end_nodes.clone().into_iter().filter_map(move |it| {
+                    let start_range = context.byte_range();
+                    let end = it.start_byte();
+                    if start_range.contains(&end)
+                            && start_range.contains(&top_first_byte)
+                            && start_range.contains(&last_scan_byte)
+                            // only match @context.end nodes that aren't at the end of the line
+                            && context.start_position().row != it.start_position().row
+                    {
+                        Some(context.start_byte()..it.start_byte().saturating_sub(1))
+                    } else {
+                        None
+                    }
+                })
+                // in some cases, the start byte of a block is on the next line
+                // which causes to show the actual first line of content instead of
+                // the actual wanted "end of signature" line
+            })
+            .next()
```

Comment:
That whole function can be rewritten as:

```rust
        // Only allocates the vector if we find at least one context with a relevant range
        let end_nodes = once_cell::unsync::Lazy::new(|| {
            query_match
                .nodes_for_capture_index(end_index)
                .collect::<Vec<_>>()
        });

        query_match
            .nodes_for_capture_index(start_index)
            .find_map(|context| {
                let ctx_start_range = context.byte_range();

                let start_range_contains_bytes = ctx_start_range.contains(&top_first_byte)
                    && ctx_start_range.contains(&last_scan_byte);
                if !start_range_contains_bytes {
                    return None;
                }

                let ctx_start_row = context.start_position().row;
                let ctx_start_byte = ctx_start_range.start;

                end_nodes.iter().find_map(|it| {
                    let end = it.start_byte();
                    // only match @context.end nodes that aren't at the end of the line
                    (ctx_start_row != it.start_position().row && ctx_start_range.contains(&end))
                        .then_some(ctx_start_byte..end.saturating_sub(1))
                })
                // in some cases, the start byte of a block is on the next line
                // which causes to show the actual first line of content instead of
                // the actual wanted "end of signature" line
            })
```

I made the commit here so you can see my process: https://github.com/poliorcetics/helix/tree/ab/sticky-nodes

I removed the clone, hoisted some checks out of the inner loop to help with early return and only allocates the `end_nodes` vec as needed to make its overall lifetime shorter (it will make invalid matches at the start much less costly to check for since there won't have been an allocation then)

Reply:
I may be mistaken but I think I kept the exact same functionalities you had, but you would know better so please check I haven't introduced a bug in here 😅

Reply:
Thanks for your work! However, i think i will rework the `context.end` in general, i think the method as of how i do it right now is quite hacky.

I will however keep this in my mind when i am doing the rework, also i think quite a lot of things, especially with iterators, are not really readable, so i want to improve that! :)